File: objectModel/Java/objectmodel/src/main/java/com/microsoft/commondatamodel/objectmodel/persistence/cdmfolder/AttributeGroupReferencePersistence.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

package com.microsoft.commondatamodel.objectmodel.persistence.cdmfolder;

import com.fasterxml.jackson.databind.JsonNode;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmAttributeGroupReference;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmCorpusContext;
import com.microsoft.commondatamodel.objectmodel.enums.CdmObjectType;
import com.microsoft.commondatamodel.objectmodel.utilities.CopyOptions;
import com.microsoft.commondatamodel.objectmodel.utilities.ResolveOptions;

public class AttributeGroupReferencePersistence {

  public static CdmAttributeGroupReference fromData(final CdmCorpusContext ctx, final JsonNode obj) {
    return fromData(ctx, obj, null);
  }

  public static CdmAttributeGroupReference fromData(final CdmCorpusContext ctx, final JsonNode obj, final String entityName) {
    if (obj == null) {
      return null;
    }
    boolean simpleReference = true;
    final Object attributeGroup;
    if (obj.isValueNode()) {
      attributeGroup = obj;
    } else {
      simpleReference = false;
      if (obj.get("attributeGroupReference").isValueNode() && obj.get("attributeGroupReference") != null) {
        attributeGroup = obj.get("attributeGroupReference").asText();
      } else {
        attributeGroup = AttributeGroupPersistence.fromData(ctx, obj.get("attributeGroupReference"), entityName);
      }
    }

    return ctx.getCorpus().makeRef(CdmObjectType.AttributeGroupRef,
            attributeGroup, simpleReference);
  }

  public static Object toData(final CdmAttributeGroupReference instance, final ResolveOptions resOpt, final CopyOptions options) {
    return CdmObjectRefPersistence.toData(instance, resOpt, options);
  }
}

File: objectModel/CSharp/Microsoft.CommonDataModel.ObjectModel/Persistence/CdmFolder/AttributeGroupReferencePersistence.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

namespace Microsoft.CommonDataModel.ObjectModel.Persistence.CdmFolder
{
    using Microsoft.CommonDataModel.ObjectModel.Cdm;
    using Microsoft.CommonDataModel.ObjectModel.Enums;
    using Microsoft.CommonDataModel.ObjectModel.Utilities;
    using Newtonsoft.Json.Linq;
    using System.Collections.Generic;

    class AttributeGroupReferencePersistence
    {
        public static CdmAttributeGroupReference FromData(CdmCorpusContext ctx, JToken obj, string entityName = null)
        {
            if (obj == null)
            {
                return null;
            }
            bool simpleReference = true;
            dynamic attributeGroup;
            if (obj is JValue)
                attributeGroup = obj;
            else
            {
                simpleReference = false;
                if (obj["attributeGroupReference"] is JValue)
                    attributeGroup = (string)obj["attributeGroupReference"];
                else
                    attributeGroup = AttributeGroupPersistence.FromData(ctx, obj["attributeGroupReference"], entityName);
            }

            CdmAttributeGroupReference attGroupReference = ctx.Corpus.MakeRef<CdmAttributeGroupReference>(CdmObjectType.AttributeGroupRef, attributeGroup, simpleReference);

            // now with applied traits!
            List<CdmTraitReferenceBase> appliedTraits = null;
            if (!(obj is JValue))
                appliedTraits = Utils.CreateTraitReferenceList(ctx, obj["appliedTraits"]);

            Utils.AddListToCdmCollection(attGroupReference.AppliedTraits, appliedTraits);

            return attGroupReference;
        }
        public static dynamic ToData(CdmAttributeGroupReference instance, ResolveOptions resOpt, CopyOptions options)
        {
            return CdmObjectRefPersistence.ToData(instance, resOpt, options);
        }

    }
}
